src/components/UploadFileDialog.tsx

```diff
@@ -0,0 +1,147 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import * as React from "react";
+import { Service } from "../service";
+import * as ReactModal from "react-modal";
+import { Button } from "./Button";
+import { GoGear, GoFile, GoX, Icon, GoPencil } from "./Icons";
+import {File, FileType, Directory, extensionForFileType, nameForFileType, filetypeForExtension} from "../model";
+import { KeyboardEvent, ChangeEvent, ChangeEventHandler } from "react";
+import { ListBox, ListItem, FileUploadInput } from "./Widgets";
+
+export interface UploadFileDialogProps {
+  isOpen: boolean;
+  directory: Directory;
+  onUpload: (file: File[]) => void;
+  onCancel: () => void;
+}
+export class UploadFileDialog extends React.Component<UploadFileDialogProps, {
+    files: Object[];
+  }> {
+  constructor(props: any) {
+    super(props);
+    this.state = {
+      files: []
+    };
+  }
+  _nameError() {
+    const directory = this.props.directory;
+    let errorStr: string = "";
+    this.state.files.map((file: any) => {
+      if (file.fileType == null) {
+          errorStr += `File '${file.name}' is not supported.`;
+      }
+      if (file.name) {
+        if (!/^[a-z0-9\.\-\_]+$/i.test(file.name)) {
+            errorStr += "Illegal characters in file name.";
+        } else if (!file.name.endsWith(extensionForFileType(file.fileType))) {
+            errorStr +=  nameForFileType(file.fileType) + " file extension is missing.";
+        } else if (directory && directory.getImmediateChild(file.name)) {
+            errorStr += `File '${file.name}' already exists.`;
+        }
+      }
+    });
+    return errorStr;
+  }
+  async _handleUpload(files: FileList) {
+    this.setState({files: []});
+    Array.from(files).forEach(async (file: any) => {
+      const name: string = file.name;
+      const fileType: string = filetypeForExtension(name.split(".").pop());
+      let data: any;
+      try {
+        data = await this._readUploadedFileAsText(file);
+        this.setState(prevState => ({
+          files: [...prevState.files, {name, fileType, data}]

```

Comment:
`data `of file is added in state even if _nameError function gives some errors. So data must be added later. I need solution for adding data in state later.